core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -203,17 +204,19 @@ public static State fromString(String state) {
     private final HashFunction routingHashFunction;
     private final boolean useTypeForRouting;
 
-    private IndexMetaData(String index, long version, State state, Settings settings, ImmutableOpenMap<String, MappingMetaData> mappings, ImmutableOpenMap<String, AliasMetaData> aliases, ImmutableOpenMap<String, Custom> customs) {
+    private IndexMetaData(String index, long version, int[] primaryTerms, State state, Settings settings, ImmutableOpenMap<String, MappingMetaData> mappings, ImmutableOpenMap<String, AliasMetaData> aliases, ImmutableOpenMap<String, Custom> customs) {
         if (settings.getAsInt(SETTING_NUMBER_OF_SHARDS, null) == null) {
             throw new IllegalArgumentException("must specify numberOfShards for index [" + index + "]");
         }
         if (settings.getAsInt(SETTING_NUMBER_OF_REPLICAS, null) == null) {
             throw new IllegalArgumentException("must specify numberOfReplicas for index [" + index + "]");
         }
+        this.settings = settings;
         this.index = index;
         this.version = version;
+        this.primaryTerms = primaryTerms;
+        assert primaryTerms.length == numberOfShards();
```

Comment:
how == nigh (at least in mind mind). We don't call the number of shards methods very frequently